ksqldb-rest-app/src/main/java/io/confluent/ksql/api/impl/BlockingPrintPublisher.java

```diff
@@ -0,0 +1,206 @@
+      final PrintTopic printTopic,
+      final Duration disconnectCheckInterval) {
+    super(ctx);
+    this.workerExecutor = Objects.requireNonNull(workerExecutor);
+    this.printTopic = printTopic;
+    this.limit = printTopic.getLimit().orElse(Integer.MAX_VALUE);
+    this.interval = printTopic.getIntervalValue();
+    this.formatter = new RecordFormatter(
+        serviceContext.getSchemaRegistryClient(),
+        printTopic.getTopic()
+    );
+    this.disconnectCheckInterval = disconnectCheckInterval;
+    this.topicConsumer = createTopicConsumer(serviceContext, consumerProperties, printTopic);
+  }
+
+  public Future<Void> close() {
+    if (closed) {
+      return Future.succeededFuture();
+    }
+    closed = true;
+    // Run async as it can block
+    ctx.runOnContext(v -> sendComplete());
+    executeOnWorker(topicConsumer::close);
+
+    return super.close();
+  }
+
+
+  @Override
+  protected void maybeSend() {
+    ctx.runOnContext(v -> {
+      ctx.executeBlocking(t -> doSend());
+    });
+  }
+
+  @Override
+  protected void afterSubscribe() {
+    if (!started) {
+      started = true;
+    }
+  }
+
+  public void startFromWorkerThread() {
+    VertxUtils.checkIsWorker();
+    started = true;
+  }
+
+  private void executeOnWorker(final Runnable runnable) {
+    workerExecutor.executeBlocking(p -> runnable.run(), false, ar -> {
+      if (ar.failed()) {
+        log.error("Failed to close print", ar.cause());
+      }
+    });
+  }
+
+  private void doSend() {
+    int messagesWritten = 0;
+    int messagesPolled = 0;
+
+    while (!isClosed() && getDemand() > 0 && messagesWritten < limit) {
+      final ConsumerRecords<Bytes, Bytes> records = topicConsumer.poll(disconnectCheckInterval);
+      if (records.isEmpty()) {
+        continue;
+      }
+
+      final List<String> formattedRecords = formatter.format(
+          records.records(this.printTopic.getTopic())
+      );
+      final List<String> formattedRecordsOld = formatConsumerRecords(records);
+
+      for (final String message : formattedRecords) {
+        if (messagesPolled++ % interval == 0) {
+          messagesWritten++;
+          doOnNext(message);
+        }
+
+        if (messagesWritten >= limit) {
+          ctx.runOnContext(v -> sendComplete());
+          break;
+        } else if (messagesWritten >= SEND_MAX_BATCH_SIZE) {
+          ctx.runOnContext(v -> doSend());
+          break;
+        }
+      }
+    }
+  }
+
+  private KafkaConsumer<Bytes, Bytes> createTopicConsumer(final ServiceContext serviceContext,
+      final Map<String, Object> consumerProperties,
+      final PrintTopic printTopic) {
+    try {
+      boolean topicExists = serviceContext.getAdminClient().listTopics().names().get().stream()
+          .anyMatch(topicName -> topicName.equalsIgnoreCase(printTopic.getTopic()));
+
+      if (!topicExists) {
+        throw new KsqlApiException("KsqlException Topic does not exist: " + printTopic.getTopic(),
```

Comment:
How come "KsqlException" is part of the exception message here?

Reply:
removed